Write explicit meta-kernels for `normal`

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70089
* #70087
* #70086
* #70085

See #69386.

Differential Revision: [D34089964](https://our.internmc.facebook.com/intern/diff/D34089964)

nkaretnikov committed:
Port tensor variants of `normal` to structured kernel

See #69386.

[ghstack-poisoned]

nkaretnikov commented:
tests/diff are [here](https://github.com/pytorch/pytorch/issues/69386#issuecomment-985931536), but i encourage you do your own testing first (in case my tests are broken).

i tried really hard to avoid BC issues and just touching public functions in general.

for tensor variants, this is the only way i've come up with that doesn't break BC, but the error messages are different in some cases. take a look at the diff mentioned above for more info.

Lezcano commented:
I'd say changing the error messages to more consistent ones is not only fine but desired.

On this topic, I think we should deprecate that `else` branch in `resize_output_for_normal` and take this chance to simplify it to be the shorter and more correct
```c++
auto shape = at::infer_size(mean.sizes(), std.sizes());
at::native::resize_output(ret, shape);
```
Even more, given that this code is so short, I would delete `resize_output_for_normal` all together and simply use this code (or even writing it as a one-liner) wherever it's needed.

What do you think @bdhirsh ?

Lezcano commented:
Also, just to make sure @nkaretnikov, does broadcasting work as expected now? If so, could you also add a sample that tests this in the OpInfo?

ysiraichi commented:
Hi @nkaretnikov. Just a small comment: (nit) maybe we should change the title of this PR, since you are not really porting `normal` variants to structured kernels.

nkaretnikov commented:
waiting for input from @bdhirsh (see above). then i'll fix the mentioned issues.

bdhirsh commented on aten/src/ATen/native/Distributions.cpp:
```diff
@@ -266,30 +267,75 @@ Tensor& normal_meta_(Tensor& self, double mean, double std, c10::optional<Genera
 
+Tensor normal_meta(const Tensor& mean, const Tensor& std, c10::optional<Generator> gen) {
+  CHECK_NORMAL_TENSOR_STD(std);
+  auto ret = at::empty_like(mean, MemoryFormat::Contiguous);
```
You might want to explicitly pass the meta device `(c10::kMeta)` into `empty_like()` here, just to be safe.

Technically, if you have mixed device inputs (say `mean` is on CPU, and `std` is on Meta), then we'll end up dispatching to this meta kernel. That happens to be true because "Meta" has higher priority than "CPU" in the dispatcher.

nkaretnikov replied:
just a comment for whoever find this: this was attempted at first, but after refactoring it's not possible to differentiate between meta/non-meta tensors anymore. so marking as resolved.

bdhirsh commented on aten/src/ATen/native/Distributions.cpp:
```diff
@@ -266,30 +267,75 @@ Tensor& normal_meta_(Tensor& self, double mean, double std, c10::optional<Genera
+
+// out float tensor
 Tensor& normal_out(double mean, const Tensor& std, c10::optional<Generator> gen, Tensor& output) {
   return at::native::templates::normal_out_impl<NormalStub, Generator>(output, mean, std, gen);
```
What do you think of doing something similar to what we do for the `uniform` meta kernels ([here](https://github.com/pytorch/pytorch/blob/adaf383837238135e93bf456ea4d6a35e7c09f69/aten/src/ATen/native/Distributions.cpp#L248))? In there, we have a (no-op) `UniformMeta` stub that we pass into the template, which lets us re-use all of the error-checking logic in the template for free. Something like:
```
return at::native::templates::normal_out_impl<NormalMeta, Generator>(output, mean, std, gen);
```

That would be kinda nice since it would make each of these functions one-liners (you don't need to construct / resize the meta tensor outputs yourself). But let me know if you think I missed something!

nkaretnikov replied:
yeah, that looks interesting. thx! i'll try to do that when i'm working on this again. i'll let you know if there are any issues

bdhirsh replied:
Hmm I don't see any changes to the PR. Did you end up going with the `return at::native::templates::normal_out_impl<NormalMeta, Generator>(output, mean, std, gen);` / did the PR get updated?

bdhirsh replied:
Oh, so this PR adds a version of meta kernels that doesn't use the `NormalMeta` template logic, but a PR further up changes the kernels added in this PR to use that? Okay, let me look more through the stack

nkaretnikov replied:
this is done now

nkaretnikov committed:
Update on "Port tensor variants of `normal` to structured kernel"


See #69386.

[ghstack-poisoned]

nkaretnikov commented:
i think i've addressed all the issues and updated the stack. pls lmk if anything is weird/broken, thx!

nkaretnikov committed:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

[ghstack-poisoned]

nkaretnikov committed:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

[ghstack-poisoned]

nkaretnikov commented:
@bdhirsh i've squashed the commits that were on top of this. please review this one, it should have everything. hope this makes it easier to review. apologies for the confusion!

bdhirsh commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -157,47 +158,6 @@ at::Tensor& random_from_to_impl(at::Tensor& self, int64_t from, c10::optional<in
-
-  if (expandable) {
-    auto shape = at::infer_size(mean.sizes(), std.sizes());
-    TORCH_CHECK(
```
Are we missing this? If i'm understanding correctly, this function currently will error out if you pass in an `output` tensor with the "wrong" size. I don't think that'll happen below when you call `resize_output()`

nkaretnikov replied:
yeah, but isn't it the whole point of this pr and resize_output? this is an intermediate step toward having meta-aware kernels for this op. and with meta you kinda always have the right size by definition (that's just the point of meta kernels -- to set up the output size). or am i missing something?

nkaretnikov replied:
discussed below, marking as resolved

bdhirsh commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -249,39 +215,40 @@ Tensor& normal_out_impl(Tensor& output, double mean, const Tensor& std, c10::opt
+  CHECK_NORMAL_TENSOR_STD(std);
+  Tensor ret = at::empty_like(mean, MemoryFormat::Contiguous);
+  auto shape = at::infer_size(mean.sizes(), std.sizes());
+  at::native::resize_output(ret, shape);
```
We can just construct `ret` with the correct shape, instead of constructing it here and resizing. You can do that with `at::empty(shape,  mean.options(), MemoryFormat::Contiguous);

nkaretnikov replied:
sure, i guess that might be faster but then the code will look more ad-hoc (i'm not sure if it affects the behavior either). so i'm in favor of keeping things as is unless it's terribly wrong

nkaretnikov replied:
discussed this with mario, it was suggested that i use your version due to potential performance implications

nkaretnikov replied:
as for other comments related to resize, i stand by what i wrote and welcome you to clarify if necessary

bdhirsh replied:
yeah perf was what I had in mind :) sounds good

nkaretnikov replied:
pushed a change that does this. pls take a look once ci succeeds

bdhirsh commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -235,8 +199,10 @@ Tensor& normal_out_impl(Tensor& output, const Tensor& mean, double std, c10::opt
-  normal_impl_<normal_kernel, RNG>(output, 0, 1, gen);
   auto mean_tensor = at::full({}, mean, output.options());
+  auto shape = at::infer_size(mean_tensor.sizes(), std.sizes());
+  at::native::resize_output(output, shape);
```
I might be wrong - do we actually need to change this function?

This looks like a behavior change. If I run this on master:
```
>>> import torch
>>> output = torch.tensor([])
>>> std = torch.ones(4)
>>> torch.normal(3, std)
tensor([2.2288, 2.9681, 2.4274, 2.2930])
>>> torch.normal(3, std, out=output)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The size of tensor a (0) must match the size of tensor b (4) at non-singleton dimension 0
```

So it looks like today, passing a size=0 `out=` tensor will raise an error instead of resizing the tensor. We probably want to keep that behavior.

nkaretnikov replied:
i don't think the current behavior is correct because in the future out with size 0 will be resized, no? so this is the step in the right direction imo. but feel free to clarify

bdhirsh replied:
ok yep, agreed. When I was originally reviewing the PR I was leaning towards "try to avoid making functionality changes to the operator when we add a meta API". But you're right, this seems like strictly better behavior (we used to incorrectly error, now we do the right thing).

bdhirsh reviewed (approved):
LGTM!

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34089964).

nkaretnikov committed:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

Differential Revision: [D34089964](https://our.internmc.facebook.com/intern/diff/D34089964)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34089964).

bdhirsh commented:
Looks like CI was red when this PR last ran - do you mind rebasing and re-running?

nkaretnikov commented:
@bdhirsh yeah, will do, sorry about the delay! i will let you know

nkaretnikov committed:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

Differential Revision: [D34089964](https://our.internmc.facebook.com/intern/diff/D34089964)

[ghstack-poisoned]

nkaretnikov commented:
@bdhirsh rebased, please take a look once ci is done

nkaretnikov committed:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

Differential Revision: [D34089964](https://our.internmc.facebook.com/intern/diff/D34089964)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34089964).

nkaretnikov commented:
just noticed there are test failures and this is still not merged. these look not relevant, i'll try repro'ing/rebasing

Commit:
Update on "Write explicit meta-kernels for `normal`"


See #69386.

Differential Revision: [D34089964](https://our.internmc.facebook.com/intern/diff/D34089964)

[ghstack-poisoned]

nkaretnikov commented:
rebased on viable/strict where the ci issues should be fixed

nkaretnikov commented:
@bdhirsh ci is green now for this stack, please take another look and merge if okay, ty!
